toucan_connectors/google_big_query/google_big_query_connector.py

```diff
@@ -226,21 +232,109 @@ def _format_columns(x: str):
 
     @staticmethod
     def _build_dataset_info_query_for_ds(dataset_id: str, db_name: str | None) -> str:
-        output = f"""
-SELECT C.table_name AS name, C.table_schema AS schema, T.table_catalog AS database,
-T.table_type AS type, C.column_name, C.data_type FROM {dataset_id}.INFORMATION_SCHEMA.COLUMNS C
-JOIN {dataset_id}.INFORMATION_SCHEMA.TABLES T ON C.table_name = T.table_name
-WHERE IS_SYSTEM_DEFINED='NO' AND IS_PARTITIONING_COLUMN='NO' AND IS_HIDDEN='NO'
+        query = f"""
+SELECT
+    C.table_name AS name,
+    C.table_schema AS schema,
+    T.table_catalog AS database,
+    T.table_type AS type,
+    C.column_name,
+    C.data_type
+FROM
+    {dataset_id}.INFORMATION_SCHEMA.COLUMNS C
+    JOIN {dataset_id}.INFORMATION_SCHEMA.TABLES T
+        ON C.table_name = T.table_name
+WHERE
+    IS_SYSTEM_DEFINED = 'NO'
+    AND IS_PARTITIONING_COLUMN = 'NO'
+    AND IS_HIDDEN = 'NO'
 """
+
         if db_name is not None:
-            output += f"AND T.table_catalog = '{db_name}'\n"
-        return output
+            query += f"AND T.table_catalog = '{db_name}'\n"
+
+        return query
 
     def _build_dataset_info_query(self, dataset_ids: Iterable[str], db_name: str | None) -> str:
         return '\nUNION ALL\n'.join(
             self._build_dataset_info_query_for_ds(dataset_id, db_name) for dataset_id in dataset_ids
         )
 
+    def _fetch_query_results_per_start_index(
+        self,
+        query: str,
+        client: bigquery.Client,
+        start_index: int,
+        query_job: QueryJob,
+        results: list[Any],
+        lock: threading.Lock,
+    ) -> None:
+        """Fetch query results for a specific start index and append to the shared results list."""
+        query_config = bigquery.QueryJobConfig()
+        query_config.query_parameters = [
+            bigquery.ScalarQueryParameter('startIndex', 'INT64', start_index)
+        ]
+
+        # Create a new query job with the updated configuration
+        query_job = client.query(query, job_config=query_config)
+
+        df_list = [df for df in self._fetch_query_results(query_job)]
+
+        # Lock the results list before appending to avoid race conditions
+        lock.acquire()
+        try:
+            results.append(df_list)
+        finally:
+            lock.release()
+
+    def _merge_query_results(self, query: str, client: bigquery.Client, num_threads: int = 6):
+        """Execute query with multiple threads, incrementing startIndex, and merge results."""
+        lock = threading.Lock()
+
+        results, threads = [], []
+
+        # Create the query job with the initial startIndex
+        # Get the total rows count
+        query_job = client.query(query)
+        total_rows = getattr(query_job, 'total_rows', _MAXIMUM_RESULTS_FETCHED)
```

Comment:
Is this always present ? If not, we might end up with a few "useless" threads, if we go with 2000 / N